Restore User announce post body & first login UI fix

willy-b opened the pull request:
- restore User announce post body
- fix "no content" popping up before signup wizard
- do not switch to '#/setup' during test
- all tests passing

Commit:
FIRST VISIT: Prevent page from flashing 'oops no content' before showing Setup

Commit:
Restore User announcement post with markdown body; make tests pass, also skip switch to '#/setup' for first login if testing

Commit:
merge

Commit:
Merged: updated test/msg-auth.js constants/hashes to reflect Ryan's magic hash changes applied to new content

willy-b commented:
Merged magic hash changes, ready for merge to master.

Commit:
Fix whitespace issues and trigger content update after initial setup

ryanxcharles commented on react/setup-modal.jsx:
```diff
@@ -69,9 +69,11 @@ let SetupModal = React.createClass({
+      let userNameSynced = dattcore.asyncSetUserName(newUserName) // *intentionally* NOT yield-ing
       window.location.hash = '#/frontpage'
+      yield userNameSynced
+      dattcore.emit('refresh-content', null)
```
Only dattcore should be emitting events on the dattcore object. If the UI needs its own events, then it needs a separate system for that.

ryanxcharles replied:
And, on that note, if you find you need to emit events in the UI, this may be time to use Flux, which I believe was invented for that purpose: https://facebook.github.io/flux/docs/overview.html

ryanxcharles commented on react/content-list.jsx:
```diff
@@ -36,7 +42,14 @@ let ContentList = React.createClass({
+      if (waitOnUserSetup) {
+        setTimeout(function () {
+          this.setState({waitOnUserSetup: false})
+        }.bind(this), 2000)
```
This seems a little odd - why two seconds? Surely there is some other way to check this other than a delay? e.g., "If no content & user has not set up yet, do not display 'no content' message".

ryanxcharles commented:
Great, Willy. Some issues:
* The tests don't pass - there are "js standard" issues. Run "npm test".
* Your commit titles are too long. Commit titles should be limited to 50 chars. If you need to write more, put a blank line after the title, and put the rest of the message below.
* See the two inline comments.

Commit:
Sep. UI events, notify peers of new user, fix test

- use EventEmitter uiEvents instead of pushing UI info on dattcore (iterating towards Flux)
- ensure UI is updated on setup modal close, etc, with "setup-closed" and "refresh-content" events
- ensure JS Standard compliance
- peers are notified when new user is announced

willy-b commented:
Separated UI events to UI specific EventEmitter (iterating towards Flux), and fixed tests.
Also got rid of 2 second delay for UI refresh and replaced it with an event-based solution. Let me know if it needs anything else for merge!

Commit:
merge

ryanxcharles commented on core/index.js:
```diff
@@ -281,9 +282,15 @@ DattCore.prototype.asyncNewContentAuth = function (title, label, body) {
+  }, this)
+}
+
+DattCore.prototype.asyncPostMsgAuth = function (msgauth) {
```
This method should have at least one test. Notice how the other tests in dattcore mock up things like broadcastMsg.

willy-b replied:
Sounds good. Also need to add a test for CoreContent#asyncGetContentAuth while I'm at it.

ryanxcharles commented on react/content-list.jsx:
```diff
@@ -45,14 +58,30 @@ let ContentList = React.createClass({
+    uiEvents.on('refresh-content', this.handleRefreshContent)
+  },
+
+  unmonitorDattCore: function () {
```
What's the reason to have an "unmonitor" function?

willy-b replied:
Good question. It is called in componentWillUnmount to remove event listeners that refer to the unmounted component. 
Not removing event listeners causes setState warnings like the following:
```
Warning: setState(...): Can only update a mounted or mounting component. This usually means you called setState() on an unmounted component. This is a no-op.
```
These still pop up in the logs but a lot less after removing content-list's dattcore listeners whenever it is unmounted via "unmonitorDattCore"

ryanxcharles replied:
I see. Then that's fine for now, but this approach is prone to error. It would be really easy to add a listener and forget to add the unlistener. Probably the way this should really be handled (not necessarily right now) is for the UI to have one high-level place that does all the listening to dattcore events, which then filter down to all the components. Individual components would then not listen to dattcore events and not even use dattcore directly.

ryanxcharles commented on react/layout.jsx:
```diff
@@ -45,13 +54,15 @@ let Layout = React.createClass({
         let userSetupFlag = yield dattcore.asyncGetUserSetupFlag()
-        if (!userSetupFlag && (window.location.hash === '#/frontpage' || !window.location.hash || window.location.hash === '#/')) {
+
+        if (!userSetupFlag && !this.state.skipSetup && (window.location.hash === '#/frontpage' || !window.location.hash || window.location.hash === '#/') && (window.location.toString().indexOf('tests.html') === -1)) {
```
This is probably fine, but I'm uneasy about using "window" for routing, because it doesn't work in node.js. The reason we would want the UI to work in node.js is so we can render it server-side before sending it. If there is a better way to handle routing that doesn't involve "window" we should use that.

ryanxcharles replied:
Another thought occurs to me about "window" - this means this file can't be easily tested in node.js either. It's very convenient for files to be able to be tested in node.js since the tests run faster that way. The UI should really be tested as thoroughly as dattcore - the only reason why it wasn't it because I didn't know a good way to test it, since I am inexperienced with React.

ryanxcharles commented on react/content-list.jsx:
```diff
@@ -45,14 +58,30 @@ let ContentList = React.createClass({
+
     dattcore.on('peers-content-auth', this.handlePeersContentAuth)
     dattcore.on('content-content-auth', this.handleContentContentAuth)
+    uiEvents.on('refresh-content', this.handleRefreshContent)
```
Much better :) But also for the record, as you mentioned somewhere else, we should probably use Flux - people complain about creating a mess by using events+react, so we should probably just adopt Flux.

ryanxcharles replied:
Since making this comment, I've looked into it a bit, and redux seems to be a good tool to use for this instead of flux. Redux is also what's used by "MERN", a tool I would like to use to replace our landing page and blog. So using Redux in datt itself would also be good for consistency.

Commit:
Refactoring and tests

- Test: DattCore#asyncPostMsgAuth confirm peer message is broadcast, that contentauth is posted, and can be retrieved via hashbuf using DattCore#asyncGetContentAuth

- Test: CoreContent#asyncGetContentAuth confirm ContentAuth hex/fields and associated Content fields match original after retrieval
- Method: DattCore#asyncGetContentAuth so UI code is not accessing dattcore.corecontent directly
- Test: DattCore#asyncGetContentAuth confirms ContentAuth hex matches original ContentAuth

- Refactor: DattCore#asyncInitialize factored into #asyncInitializeDb, #asyncInitializeCoreBitcoin, #asyncInitializeCoreContent, etc, to make it easier to partially initialize DattCore for tests and for readability
- Method: DattCore#whenInitialized. Returns a Promise resolved if/when DattCore and all dependencies are initialized

Commit:
merge

ryanxcharles commented on core/index.js:
```diff
@@ -69,31 +69,83 @@ DattCore.prototype.initialize = function () {
 /**
+ * Get a promise resolved if/when DattCore and all its dependencies are finished initializing
+ */
+DattCore.prototype.whenInitialized = function () {
```
This should be prefixed with the word "async".

ryanxcharles commented on core/index.js:
```diff
@@ -69,31 +69,83 @@ DattCore.prototype.initialize = function () {
+        resolve(this.isinitialized)
+      }.bind(this))
+    }
+  }.bind(this))
```
Rather than using "bind" here, you could just use arrow functions.

ryanxcharles commented on react/content-list.jsx:
```diff
@@ -17,11 +17,18 @@ let ContentList = React.createClass({
   },
 
-  setStateFromDattCore: function () {
+  setStateFromDattCore: function (skipSetup) {
```
As mentioned previously, IMO, the "setup" simply shouldn't exist at all. I don't think we gain anything by asking the user to "setup" their account. It removes one of the great usability benefits of a decentralized system: zero setup.

willy-b replied:
Agree. However, I still think we should just complete Chris' mockups before taking a step back and revising/iterating. 
This will keep us honest about our milestone and also provide usecase validation for the Datt core methods by having a UI to work with.

ryanxcharles commented:
Comments that don't prevent a merge:
* Do not merge from master, especially not more than once. This will create a nonlinear history that is difficult to understand. You should be rebasing your PRs on top of master. The only merge commits that exist should be merging into master.
* We should be using Redux instead of events. After inquiring with @maraoz, he agrees and argues Redux is better than events :)

ryanxcharles commented on core/index.js:
```diff
@@ -69,31 +69,83 @@ DattCore.prototype.initialize = function () {
+
+    this.isinitialized = true
+    this.emit('initialized', this.isinitialized)
+    return Promise.resolve()
```
Inside asink you can just return, and it will return a promise automatically. You don't need to return Promise.resolve(). This is the same as with the async/await pattern.

ryanxcharles commented:
Also, your commit messages should be no wider than 80 chars. Your editor should have a way to limit the width to 80 chars. See how this looks non-wrapped when viewing with default settings with 'git log':

![screenshot from 2016-02-23 14 00 50](https://cloud.githubusercontent.com/assets/63740/13268210/123f3dd8-da36-11e5-8b43-ca1adf511e71.png)

ryanxcharles commented:
I get a test failure when running the tests. This prevents a merge.

<pre>
  208 passing (15s)
  1 pending
  1 failing

  1) Layout "after all" hook:
     TypeError: Cannot read property 'unmonitorBlockchainAPI' of undefined
      at DattCore.<anonymous> (index.js:177:5)
      at [object Generator].next (native)
      at node_modules/asink/index.js:47:35
      at step (node_modules/asink/index.js:29:16)
      at node_modules/asink/index.js:47:5
      at new Promise (node_modules/karma/node_modules/core-js/modules/es6.promise.js:209:7)
      at spawn (node_modules/asink/index.js:24:10)
      at DattCore.asyncNetworkClose (index.js:176:10)
      at layout.jsx:28:22
      at [object Generator].next (native)
      at node_modules/asink/index.js:47:35
      at step (node_modules/asink/index.js:29:16)
      at node_modules/asink/index.js:47:5
      at new Promise (node_modules/karma/node_modules/core-js/modules/es6.promise.js:209:7)
      at spawn (node_modules/asink/index.js:24:10)
      at Context.<anonymous> (layout.jsx:27:12)
</pre>

Commit:
Fix a couple of undefined and null handling bugs and add test

- Fix "TypeError: Cannot read property unmonitorBlockchainAPI of undefined"
- Fix BlockchainAPI#asyncGetRequest handling of null response (e.g. if disconnected)
- Test - DattCore#monitorCoreBitcoin: verify that event handlers for "balance" and "block-info" are set

ryanxcharles commented:
Project has been rearchitected - PR no longer applies.
